internal/clients/http/rbac/rbac_client.go

```diff
@@ -76,57 +79,74 @@ func (c *rbac) GetPrincipalAccess(ctx context.Context) (clients.RbacAcl, error)
-	maxQueries := 10
-
-	// keep fetching until we have all the records
-	for len(result) < records {
-		params := GetPrincipalAccessParams{
-			Application: "provisioning",
-			Limit:       FetchLimit,
-			Offset:      &offset,
-		}
-		resp, err := c.client.GetPrincipalAccessWithResponse(ctx, &params, headers.AddRbacIdentityHeader, headers.AddEdgeRequestIdHeader)
-		if err != nil {
-			return nil, fmt.Errorf("get principal access: %w", err)
-		}
 
-		err = http.HandleHTTPResponses(ctx, resp.StatusCode())
-		if err != nil {
-			return nil, fmt.Errorf("get principal access: %w", err)
+	err := cache.Find(ctx, orgID+accountNumber, &result)
+	if errors.Is(err, cache.ErrNotFound) {
+		fullRequest = true
+		start := time.Now()
+		defer func() {
+			metrics.RbacAclFetchDuration.Observe(float64(time.Since(start).Nanoseconds()) / 1000000)
+		}()
+
+		records := math.MaxInt
+		offset := 0
+		maxQueries := 10
+
+		// keep fetching until we have all the records
+		for len(result) < records {
+			params := GetPrincipalAccessParams{
+				Application: "provisioning",
+				Limit:       FetchLimit,
+				Offset:      &offset,
+			}
+			resp, gpaErr := c.client.GetPrincipalAccessWithResponse(ctx, &params, headers.AddRbacIdentityHeader, headers.AddEdgeRequestIdHeader)
+			if gpaErr != nil {
+				return nil, fmt.Errorf("get principal access: %w", gpaErr)
+			}
+
+			gpaErr = http.HandleHTTPResponses(ctx, resp.StatusCode())
+			if gpaErr != nil {
+				return nil, fmt.Errorf("get principal access: %w", gpaErr)
+			}
+			logger.Trace().
+				Int("limit", *FetchLimit).
+				Int("offset", offset).
+				Int("length", len(result)).
+				Int("entries", len(resp.JSON200.Data)).
+				Int("return_code", resp.StatusCode()).
+				Msg("Performed get principal access RBAC call")
+
+			if resp.JSON200 == nil || resp.JSON200.Meta == nil || resp.JSON200.Meta.Count == nil {
+				return nil, ErrMetaNotPresent
+			}
+			records = int(*resp.JSON200.Meta.Count)
+
+			for _, a := range resp.JSON200.Data {
+				result = append(result, clients.NewAccess(a.Permission))
+			}
+			offset += *FetchLimit
+
+			maxQueries -= 1
+			if maxQueries <= 0 {
+				logger.Warn().Msg("Maximum amount of RBAC requests reached, giving up")
+				break
+			}
 		}
-		zerolog.Ctx(ctx).Trace().
-			Int("limit", *FetchLimit).
-			Int("offset", offset).
-			Int("length", len(result)).
-			Int("entries", len(resp.JSON200.Data)).
-			Int("return_code", resp.StatusCode()).
-			Msg("Performed get principal access RBAC call")
-
-		if resp.JSON200 == nil || resp.JSON200.Meta == nil || resp.JSON200.Meta.Count == nil {
-			return nil, ErrMetaNotPresent
-		}
-		records = int(*resp.JSON200.Meta.Count)
-
-		for _, a := range resp.JSON200.Data {
-			result = append(result, clients.NewAccess(a.Permission))
-		}
-		offset += *FetchLimit
 
-		maxQueries -= 1
-		if maxQueries <= 0 {
-			zerolog.Ctx(ctx).Warn().Msg("Maximum amount of RBAC requests reached, giving up")
-			break
+		err = cache.Set(ctx, orgID+accountNumber, &result)
```

Comment:
How big can the result be? I se we are storing it all, so we are serializing all permissions the user has.
I see those are filtered, so it should be just permissions to our app, so I guess not a big issue, just verifying :)

Reply:
This is a slice of

```
type Access struct {
	Resource string `json:"resource"`
	Verb     string `json:"verb"`
}
```

and on our platform we have several dozens of permissions. In the worst case scenario, perhaps hundreds of elements in the slice. This gets serialized by Gob encoder and stored as blob, it cannot get faster than that.

What is your concern? The bigger the result is, the better caching actually is because transferring a binary blob to Redis will be likely faster than HTTP/JSON (maybe with pagination in the future). Once this is merged, opentelemetry will clearly answer how much faster this is (or not).

Reply:
My concern is huge documents in Redis, I've heard stories, that redis is not particularly great at handling huge documents, but given I do not have enough info to what "huge" means, it's just an unsupported worry :)
I think the main issue is that Redis is not concurrent, but not sure at all :)

Aren't we filtering the results to only include provisioning permissions tho?

> The bigger the result is, the better caching actually is because transferring

I see biggest value in saving round trip and the computation needed to return the result, I'd not be so sure that the download itself will be significantly faster for big documents (once the connection is established) :thinking:

Reply:
> My concern is huge documents in Redis

We are not storing documents, but binary blobs. Maximum value size is 512 MB of each entry, according to google.

> Aren't we filtering the results to only include provisioning permissions tho?

Actually, we only query permissions for "provisioning" so I guess this is the answer to your concern:

```
		params := GetPrincipalAccessParams{
			Application: "provisioning",
			Limit:       FetchLimit,
			Offset:      &offset,
		}
```

Average size of an entry will be less than a memory page.

Okay, enough theory, let’s merge this so I still have a day to provide a fix. Stage logs should answer us what is faster - RBAC API or Redis. It can be anything actually.